File: lobster-input-ios/lobster-input-ios/Models/APIModels.swift
import Foundation

// MARK: - 认证

struct SendCodeRequest: Encodable {
    let email: String
}

struct LoginRequest: Encodable {
    let email: String
    let code: String
    let deviceId: String
    let hardwareFingerprint: String

    enum CodingKeys: String, CodingKey {
        case email, code
        case deviceId = "device_id"
        case hardwareFingerprint = "hardware_fingerprint"
    }
}

struct AuthResponse: Decodable {
    let token: String
    let email: String
    let tier: String
    let isNewUser: Bool
    let requireInvite: Bool

    enum CodingKeys: String, CodingKey {
        case token, email, tier
        case isNewUser     = "is_new_user"
        case requireInvite = "require_invite"
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        token         = try c.decode(String.self, forKey: .token)
        email         = try c.decode(String.self, forKey: .email)
        tier          = (try? c.decode(String.self, forKey: .tier)) ?? "trial"
        isNewUser     = (try? c.decode(Bool.self, forKey: .isNewUser)) ?? false
        requireInvite = (try? c.decode(Bool.self, forKey: .requireInvite)) ?? false
    }
}

struct VerifyInviteRequest: Encodable {
    let email: String
    let inviteCode: String
    let deviceId: String
    let hardwareFingerprint: String

    enum CodingKeys: String, CodingKey {
        case email
        case inviteCode = "invite_code"
        case deviceId   = "device_id"
        case hardwareFingerprint = "hardware_fingerprint"
    }
}

struct InviteCodeItem: Decodable, Identifiable {
    var id: String { code }
    let code: String
    let isUsed: Bool
    let usedBy: String?
    let usedAt: String?

    enum CodingKeys: String, CodingKey {
        case code
        case isUsed = "is_used"
        case usedBy = "used_by"
        case usedAt = "used_at"
    }
}

struct MyInviteCodesResponse: Decodable {
    let inviteCodes: [InviteCodeItem]

    enum CodingKeys: String, CodingKey {
        case inviteCodes = "invite_codes"
    }
}

// MARK: - 套餐积分

struct UserPlanInfo: Decodable {
    let tier: String
    /// 后端按 X-Accept-Language 返回的本地化套餐名(套餐文案统一后端管理);优先展示,空则回退本地
    let planName: String
    let creditsTotal: Int
    let creditsUsed: Int
    let creditsRemaining: Int
    let creditsResetAt: String?
    /// 当前套餐到期时间（UTC ISO8601），免费/永久套餐为空
    let planExpiresAt: String?
    /// 当前订阅到期时间（UTC ISO8601），免费套餐可能为空
    let subscriptionExpiresAt: String?
    let registrationEnabled: Bool
    let inviteCodeEnabled: Bool
    let showInviteCodesEnabled: Bool
    /// 订阅是否处于自动续费状态
    let autoRenew: Bool
    /// 下次自动续费时间（UTC ISO8601），非自动续费为空
    let nextRenewalAt: String?
    /// 是否可在客户端内取消自动续费（Apple IAP 订阅后端恒为 false，取消走系统「管理订阅」）
    let renewalCancellable: Bool

    enum CodingKeys: String, CodingKey {
        case tier
        case planName               = "plan_name"
        case creditsTotal           = "credits_total"
        case creditsUsed            = "credits_used"
        case creditsRemaining       = "credits_remaining"
        case creditsResetAt         = "credits_reset_at"
        case planExpiresAt          = "plan_expires_at"
        case subscriptionExpiresAt  = "subscription_expires_at"
        case registrationEnabled    = "registration_enabled"
        case inviteCodeEnabled      = "invite_code_enabled"
        case showInviteCodesEnabled = "show_invite_codes_enabled"
        case autoRenew              = "auto_renew"
        case nextRenewalAt          = "next_renewal_at"
        case renewalCancellable     = "renewal_cancellable"
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        tier                    = (try? c.decode(String.self, forKey: .tier)) ?? "none"
        planName                = (try? c.decode(String.self, forKey: .planName)) ?? ""
        creditsTotal            = (try? c.decode(Int.self, forKey: .creditsTotal)) ?? 0
        creditsUsed             = (try? c.decode(Int.self, forKey: .creditsUsed)) ?? 0
        creditsRemaining        = (try? c.decode(Int.self, forKey: .creditsRemaining)) ?? 0
        creditsResetAt          = try? c.decode(String.self, forKey: .creditsResetAt)
        planExpiresAt           = try? c.decode(String.self, forKey: .planExpiresAt)
        subscriptionExpiresAt   = try? c.decode(String.self, forKey: .subscriptionExpiresAt)
        registrationEnabled     = (try? c.decode(Bool.self, forKey: .registrationEnabled)) ?? true
        inviteCodeEnabled       = (try? c.decode(Bool.self, forKey: .inviteCodeEnabled)) ?? false
        showInviteCodesEnabled  = (try? c.decode(Bool.self, forKey: .showInviteCodesEnabled)) ?? false
        autoRenew               = (try? c.decode(Bool.self, forKey: .autoRenew)) ?? false
        nextRenewalAt           = try? c.decode(String.self, forKey: .nextRenewalAt)
        renewalCancellable      = (try? c.decode(Bool.self, forKey: .renewalCancellable)) ?? false
    }
}

/// POST /api/v1/subscription/cancel-renewal 响应
struct CancelRenewalResponse: Decodable {
    /// "cancelled" | "already_cancelled" | "apple_managed"
    let status: String
    /// 取消后套餐仍可使用至的时间（UTC ISO8601），可能为空
    let effectiveUntil: String?

    enum CodingKeys: String, CodingKey {
        case status
        case effectiveUntil = "effective_until"
    }
}

// MARK: - App 启动配置

struct AppStartupConfig: Decodable {
    let registrationEnabled: Bool
    let inviteCodeEnabled: Bool
    let showInviteCodesEnabled: Bool
    let registrationLimitEnabled: Bool
    let registrationLimitCount: Int

    enum CodingKeys: String, CodingKey {
        case registrationEnabled     = "registration_enabled"
        case inviteCodeEnabled        = "invite_code_enabled"
        case showInviteCodesEnabled   = "show_invite_codes_enabled"
        case registrationLimitEnabled = "registration_limit_enabled"
        case registrationLimitCount   = "registration_limit_count"
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        registrationEnabled     = (try? c.decode(Bool.self, forKey: .registrationEnabled)) ?? true
        inviteCodeEnabled        = (try? c.decode(Bool.self, forKey: .inviteCodeEnabled)) ?? true
        showInviteCodesEnabled   = (try? c.decode(Bool.self, forKey: .showInviteCodesEnabled)) ?? false
        registrationLimitEnabled = (try? c.decode(Bool.self, forKey: .registrationLimitEnabled)) ?? false
        registrationLimitCount   = (try? c.decode(Int.self, forKey: .registrationLimitCount)) ?? 1000
    }
}

// MARK: - 音频处理

/// iOS 端支持的 ActionType（无 OpenClaw 系列）
enum ActionType: String, Decodable {
    case paste = "paste"
    case clarify = "clarify"
    case showMarkdown = "show_markdown"
    case tip = "tip"
}

struct ConfigUpdatePayload: Decodable {
    let maxDurationSec: Int?

    enum CodingKeys: String, CodingKey {
        case maxDurationSec = "max_duration_sec"
    }
}

struct AudioProcessResponse: Decodable {
    let operation: String
    let actionType: ActionType
    let transcript: String
    let result: String
    let modelProvider: String?
    let modelName: String?
    let warning: String?
    let configUpdate: ConfigUpdatePayload?
    let clarifyQuestion: String?
    let creditsRemaining: Int?

    enum CodingKeys: String, CodingKey {
        case operation, transcript, result, warning
        case actionType       = "action_type"
        case modelProvider    = "model_provider"
        case modelName        = "model_name"
        case configUpdate     = "config_update"
        case clarifyQuestion  = "clarify_question"
        case creditsRemaining = "credits_remaining"
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        operation       = try c.decode(String.self, forKey: .operation)
        transcript      = try c.decode(String.self, forKey: .transcript)
        result          = try c.decode(String.self, forKey: .result)
        modelProvider   = try? c.decode(String.self, forKey: .modelProvider)
        modelName       = try? c.decode(String.self, forKey: .modelName)
        warning         = try? c.decode(String.self, forKey: .warning)
        configUpdate    = try? c.decode(ConfigUpdatePayload.self, forKey: .configUpdate)
        clarifyQuestion = try? c.decode(String.self, forKey: .clarifyQuestion)
        creditsRemaining = try? c.decode(Int.self, forKey: .creditsRemaining)
        let rawAction   = (try? c.decode(String.self, forKey: .actionType)) ?? "paste"
        actionType      = ActionType(rawValue: rawAction) ?? .paste
    }
}

struct TextProcessRequest: Encodable {
    let operation: String
    let text: String
    let clientASRText: String
    let asrSessionID: String
    let selectedText: String?
    let clipboardHistory: [String]?
    let provider: String?
    let model: String?
    let openclawStatus: String?
    let fastMode: Bool
    let transcriptLanguage: String?

    enum CodingKeys: String, CodingKey {
        case operation, text
        case clientASRText = "client_asr_text"
        case asrSessionID = "asr_session_id"
        case selectedText = "selected_text"
        case clipboardHistory = "clipboard_history"
        case provider, model
        case openclawStatus = "openclaw_status"
        case fastMode = "fast_mode"
        case transcriptLanguage = "transcript_language"
    }
}

enum MobileQuickAction: String, Codable, CaseIterable {
    case format
    case polish
    case concise
    case bullets
}

struct MobileQuickActionRequest: Encodable {
    let action: MobileQuickAction
    let text: String
    let sourceOperation: String?

    enum CodingKeys: String, CodingKey {
        case action, text
        case sourceOperation = "source_operation"
    }
}

struct TextQuickActionResponse: Decodable {
    let operation: String
    let actionType: ActionType
    let inputText: String
    let result: String
    let modelProvider: String?
    let modelName: String?
    let creditsRemaining: Int?
    let transcript: String?

    enum CodingKeys: String, CodingKey {
        case operation, result, transcript
        case actionType = "action_type"
        case inputText = "input_text"
        case modelProvider = "model_provider"
        case modelName = "model_name"
        case creditsRemaining = "credits_remaining"
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        operation = try c.decode(String.self, forKey: .operation)
        inputText = try c.decode(String.self, forKey: .inputText)
        result = try c.decode(String.self, forKey: .result)
        modelProvider = try? c.decode(String.self, forKey: .modelProvider)
        modelName = try? c.decode(String.self, forKey: .modelName)
        creditsRemaining = try? c.decode(Int.self, forKey: .creditsRemaining)
        transcript = try? c.decode(String.self, forKey: .transcript)
        let rawAction = (try? c.decode(String.self, forKey: .actionType)) ?? "paste"
        actionType = ActionType(rawValue: rawAction) ?? .paste
    }
}

// MARK: - 错误处理

struct APIErrorResponse: Decodable {
    let code: String
    let message: String
    let configUpdate: ConfigUpdatePayload?

    enum CodingKeys: String, CodingKey {
        case code, message
        case configUpdate = "config_update"
    }
}

enum APIError: LocalizedError {
    case httpError(Int, String, APIErrorResponse?)
    case decodingError(Error)
    case networkError(Error)
    case unauthorized
    case userBanned
    case notLoggedIn

    var errorDescription: String? {
        switch self {
        case .httpError(_, _, let resp):
            if let code = resp?.code {
                return L10n.errorForCode(code) ?? resp?.message ?? L10n.errorUnknown
            }
            return resp?.message ?? L10n.errorUnknown
        case .decodingError: return L10n.errorDecoding
        case .networkError: return L10n.errorNetwork
        case .unauthorized: return L10n.errorUnauthorized
        case .userBanned: return L10n.errorUserBanned
        case .notLoggedIn: return L10n.errorNotLoggedIn
        }
    }

    var errorCode: String? {
        if case .httpError(_, _, let resp) = self { return resp?.code }
        return nil
    }

    var configUpdate: ConfigUpdatePayload? {
        if case .httpError(_, _, let resp) = self { return resp?.configUpdate }
        return nil
    }

    var isCreditsExhausted: Bool { errorCode == "CREDITS_EXHAUSTED" }
    var isNonRetryable: Bool { errorCode == "DURATION_EXCEEDED" }
}

struct EmptyResponse: Decodable {}

// MARK: - 录音配置

struct RecordingConfigResponse: Decodable {
    let maxDurationSec: Int

    enum CodingKeys: String, CodingKey {
        case maxDurationSec = "max_duration_sec"
    }
}

// MARK: - 词典（热词）

struct HotWordItem: Codable, Identifiable {
    let id: String
    var word: String
    let createdAt: String?

    enum CodingKeys: String, CodingKey {
        case id, word
        case createdAt = "created_at"
    }
}

struct HotWordCreateBody: Encodable { let word: String }
struct HotWordUpdateBody: Encodable { let word: String }
struct HotWordListResponse: Decodable { let hotwords: [HotWordItem] }

// MARK: - 人设配置

let personaPromptMaxLength = 4000
let personaNameMaxLength   = 60
let personaDescMaxLength   = 200
let personaMaxCount        = 10

struct PersonaPrompts: Codable {
    var transcribePrompt: String?
    var transcribeEnabled: Bool
    var rewritePrompt: String?
    var rewriteEnabled: Bool
    var intentHint: String?
    var intentEnabled: Bool

    enum CodingKeys: String, CodingKey {
        case transcribePrompt  = "transcribe_prompt"
        case transcribeEnabled = "transcribe_enabled"
        case rewritePrompt     = "rewrite_prompt"
        case rewriteEnabled    = "rewrite_enabled"
        case intentHint        = "intent_hint"
        case intentEnabled     = "intent_enabled"
    }

    init(
        transcribePrompt: String? = nil, transcribeEnabled: Bool = false,
        rewritePrompt: String? = nil, rewriteEnabled: Bool = false,
        intentHint: String? = nil, intentEnabled: Bool = false
    ) {
        self.transcribePrompt  = transcribePrompt
        self.transcribeEnabled = transcribeEnabled
        self.rewritePrompt     = rewritePrompt
        self.rewriteEnabled    = rewriteEnabled
        self.intentHint        = intentHint
        self.intentEnabled     = intentEnabled
    }
}

struct PersonaItem: Codable, Identifiable {
    let id: String
    var name: String
    var description: String?
    var isActive: Bool
    var isBuiltin: Bool
    var prompts: PersonaPrompts

    enum CodingKeys: String, CodingKey {
        case id, name, description, prompts
        case isActive = "is_active"
        case isBuiltin = "is_builtin"
    }

    init(
        id: String = UUID().uuidString, name: String = "",
        description: String? = nil, isActive: Bool = false,
        isBuiltin: Bool = false,
        prompts: PersonaPrompts = PersonaPrompts()
    ) {
        self.id = id; self.name = name; self.description = description
        self.isActive = isActive; self.isBuiltin = isBuiltin
        self.prompts = prompts
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id = try c.decode(String.self, forKey: .id)
        name = try c.decode(String.self, forKey: .name)
        description = try? c.decode(String.self, forKey: .description)
        isActive = (try? c.decode(Bool.self, forKey: .isActive)) ?? false
        isBuiltin = (try? c.decode(Bool.self, forKey: .isBuiltin)) ?? false
        prompts = (try? c.decode(PersonaPrompts.self, forKey: .prompts)) ?? PersonaPrompts()
    }
}

struct PersonaListResponse: Decodable { let personas: [PersonaItem] }
struct PersonaCreateBody: Encodable { let name: String; let description: String?; let prompts: PersonaPrompts }
struct PersonaUpdateBody: Encodable { let name: String?; let description: String?; let prompts: PersonaPrompts? }

// MARK: - 协议

struct AgreementItem: Decodable, Identifiable {
    let id: String
    let type: String
    let title: String
    let content: String
    let version: String
}

struct AgreementsResponse: Decodable {
    let agreements: [AgreementItem]
}

File: lobster-input-android/app/src/main/java/com/lobster/input/data/model/ApiModels.kt
package com.lobster.input.data.model

import android.content.Context
import com.google.gson.annotations.SerializedName
import com.lobster.input.core.locale.MobileStrings

// ==================== 认证相关 ====================

data class SendCodeRequest(
    val email: String
)

data class LoginRequest(
    val email: String,
    val code: String,
    @SerializedName("device_id")
    val deviceId: String,
    @SerializedName("hardware_fingerprint")
    val hardwareFingerprint: String = ""
)

data class AuthResponse(
    val token: String,
    val email: String,
    val tier: String = "trial",
    @SerializedName("is_new_user")
    val isNewUser: Boolean = false,
    @SerializedName("require_invite")
    val requireInvite: Boolean = false
)

data class VerifyInviteRequest(
    val email: String,
    @SerializedName("invite_code")
    val inviteCode: String,
    @SerializedName("device_id")
    val deviceId: String,
    @SerializedName("hardware_fingerprint")
    val hardwareFingerprint: String = ""
)

data class InviteCodeItem(
    val code: String,
    @SerializedName("is_used")
    val isUsed: Boolean,
    @SerializedName("used_by")
    val usedBy: String?,
    @SerializedName("used_at")
    val usedAt: String?
)

data class MyInviteCodesResponse(
    @SerializedName("invite_codes")
    val inviteCodes: List<InviteCodeItem>
)

// ==================== 套餐积分 ====================

data class UserPlanInfo(
    val tier: String = "none",
    // 后端按 X-Accept-Language 返回的本地化套餐名(套餐文案统一后端管理);优先展示,空则回退本地翻译
    @SerializedName("plan_name")
    val planName: String = "",
    @SerializedName("credits_total")
    val creditsTotal: Int = 0,
    @SerializedName("credits_used")
    val creditsUsed: Int = 0,
    @SerializedName("credits_remaining")
    val creditsRemaining: Int = 0,
    @SerializedName("credits_reset_at")
    val creditsResetAt: String? = null,
    @SerializedName("plan_expires_at")
    val planExpiresAt: String? = null,
    @SerializedName("subscription_expires_at")
    val subscriptionExpiresAt: String? = null,
    @SerializedName("registration_enabled")
    val registrationEnabled: Boolean = true,
    @SerializedName("invite_code_enabled")
    val inviteCodeEnabled: Boolean = false,
    @SerializedName("show_invite_codes_enabled")
    val showInviteCodesEnabled: Boolean = false,
    @SerializedName("auto_renew")
    val autoRenew: Boolean = false,
    @SerializedName("next_renewal_at")
    val nextRenewalAt: String? = null,
    @SerializedName("renewal_cancellable")
    val renewalCancellable: Boolean = false
)

// ==================== 订阅自动续费 ====================

/** 取消自动续费请求:后端约定空 JSON body,Gson 将无字段类序列化为 {}。 */
class CancelRenewalBody

data class CancelRenewalResponse(
    val status: String = "",
    @SerializedName("effective_until")
    val effectiveUntil: String? = null
)

// ==================== 支付与订阅 ====================

data class PaymentCatalogResponse(
    @SerializedName("active_provider")
    val activeProvider: String = "",
    val providers: List<PaymentProvider> = emptyList(),
    @SerializedName("payment_methods")
    val paymentMethods: List<PaymentMethod> = emptyList(),
    @SerializedName("subscription_module")
    val subscriptionModule: SubscriptionModuleState = SubscriptionModuleState(),
    @SerializedName("current_plan")
    val currentPlan: PaymentCurrentPlan = PaymentCurrentPlan(),
    val subscriptions: List<PaymentSubscriptionPlan> = emptyList(),
    @SerializedName("credits_topup")
    val creditsTopup: PaymentCreditsTopup = PaymentCreditsTopup(),
    val discount: PaymentDiscount = PaymentDiscount()
)

data class PaymentProvider(
    val code: String = "",
    val name: String = "",
    val status: String? = null
)

data class PaymentMethod(
    val code: String = "",
    val name: String = "",
    val description: String? = null,
    val enabled: Boolean? = null,
    @SerializedName("sort_order")
    val sortOrder: Int? = null,
    val currencies: List<String>? = null
)

data class SubscriptionModuleState(
    val enabled: Boolean = true
)

data class PaymentCurrentPlan(
    @SerializedName("plan_code")
    val planCode: String = "free",
    // 后端本地化套餐名(catalog current_plan);优先展示
    @SerializedName("plan_name")
    val planName: String = "",
    @SerializedName("plan_credits_total")
    val planCreditsTotal: Int = 0,
    @SerializedName("plan_credits_used")
    val planCreditsUsed: Int = 0,
    @SerializedName("plan_credits_remaining")
    val planCreditsRemaining: Int = 0,
    val paid: Boolean = false,
    @SerializedName("paid_topup_enabled")
    val paidTopupEnabled: Boolean = false,
    // 当前套餐的账期(monthly/yearly 等),旧后端不返回时为 null
    @SerializedName("billing_cycle")
    val billingCycle: String? = null
)

data class PaymentSubscriptionPlan(
    val type: String = "subscription",
    @SerializedName("plan_code")
    val planCode: String = "",
    val name: String = "",
    val credits: Int = 0,
    val rank: Int = 0,
    @SerializedName("billing_options")
    val billingOptions: List<PaymentBillingOption> = emptyList()
)

data class PaymentBillingOption(
    val cycle: String = "",
    @SerializedName("product_code")
    val productCode: String? = null,
    @SerializedName("price_cents")
    val priceCents: Int = 0,
    val currency: String? = null,
    @SerializedName("payment_methods")
    val paymentMethods: List<PaymentMethod>? = null,
    @SerializedName("duration_period")
    val durationPeriod: String? = null,
    @SerializedName("duration_count")
    val durationCount: Int = 1,
    // —— 以下为月付→年付补差价升级新增字段,旧后端不返回时均为 null ——
    // 实际应付金额(分),升级场景下为补差价后的金额
    @SerializedName("payable_price_cents")
    val payablePriceCents: Int? = null,
    // 结算方式:full_price(全价) / prorated_difference(按剩余价值补差价)
    @SerializedName("settlement_mode")
    val settlementMode: String? = null,
    // 升级时抵扣的未使用部分金额(分)
    @SerializedName("upgrade_credit_cents")
    val upgradeCreditCents: Int? = null,
    // 服务端权威的可购判定,null 表示旧后端未提供
    val purchasable: Boolean? = null,
    // 不可购原因:duplicate_purchase / lower_tier / cycle_downgrade
    @SerializedName("blocked_reason")
    val blockedReason: String? = null
) {
    /** 实际应付金额(分):优先取服务端下发的应付价,旧后端回退到原价。 */
    val effectivePriceCents: Int
        get() = payablePriceCents ?: priceCents

    /** 是否为补差价升级:服务端标记按剩余价值结算且应付价低于原价。 */
    val isProratedUpgrade: Boolean
        get() = settlementMode == "prorated_difference" && effectivePriceCents < priceCents
}

data class PaymentCreditsTopup(
    val type: String = "credits_topup",
    val provider: String = "",
    @SerializedName("product_code")
    val productCode: String? = null,
    val enabled: Boolean = false,
    val available: Boolean = false,
    val amount: Int? = null,
    @SerializedName("price_cents")
    val priceCents: Int? = null,
    val currency: String? = null,
    @SerializedName("payment_methods")
    val paymentMethods: List<PaymentMethod>? = null,
    @SerializedName("blocked_reason")
    val blockedReason: String? = null
)

data class PaymentDiscount(
    val supported: Boolean = true,
    @SerializedName("default_enabled")
    val defaultEnabled: Boolean = false
)

data class CreateSubscriptionCheckoutBody(
    @SerializedName("plan_code")
    val planCode: String,
    @SerializedName("billing_cycle")
    val billingCycle: String,
    @SerializedName("auto_renew")
    val autoRenew: Boolean,
    val provider: String,
    @SerializedName("product_code")
    val productCode: String,
    @SerializedName("payment_method")
    val paymentMethod: String,
    val currency: String,
    @SerializedName("settlement_mode")
    val settlementMode: String,
    @SerializedName("discount_code")
    val discountCode: String
)

data class CreateCreditsTopupCheckoutBody(
    val provider: String,
    @SerializedName("product_code")
    val productCode: String,
    @SerializedName("payment_method")
    val paymentMethod: String,
    val currency: String,
    @SerializedName("discount_code")
    val discountCode: String
)

data class PaymentCheckoutResponse(
    val provider: String = "",
    @SerializedName("request_id")
    val requestId: String? = null,
    @SerializedName("checkout_id")
    val checkoutId: String? = null,
    @SerializedName("checkout_url")
    val checkoutUrl: String = "",
    val status: String? = null
)

// ==================== App启动配置 ====================

data class AppStartupConfig(
    @SerializedName("registration_enabled")
    val registrationEnabled: Boolean = true,
    @SerializedName("invite_code_enabled")
    val inviteCodeEnabled: Boolean = true,
    @SerializedName("show_invite_codes_enabled")
    val showInviteCodesEnabled: Boolean = false,
    @SerializedName("registration_limit_enabled")
    val registrationLimitEnabled: Boolean = false,
    @SerializedName("registration_limit_count")
    val registrationLimitCount: Int = 1000
)

// ==================== 音频处理 ====================

enum class ActionType {
    @SerializedName("paste")
    PASTE,
    @SerializedName("clarify")
    CLARIFY,
    @SerializedName("show_markdown")
    SHOW_MARKDOWN,
    @SerializedName("tip")
    TIP
}

data class ConfigUpdatePayload(
    @SerializedName("max_duration_sec")
    val maxDurationSec: Int?
)

data class AudioProcessResponse(
    val operation: String,
    @SerializedName("action_type")
    val actionType: ActionType = ActionType.PASTE,
    val transcript: String,
    val result: String,
    @SerializedName("model_provider")
    val modelProvider: String? = null,
    @SerializedName("model_name")
    val modelName: String? = null,
    val warning: String? = null,
    @SerializedName("config_update")
    val configUpdate: ConfigUpdatePayload? = null,
    @SerializedName("clarify_question")
    val clarifyQuestion: String? = null,
    @SerializedName("credits_remaining")
    val creditsRemaining: Int? = null
)

data class TextProcessRequest(
    val operation: String,
    val text: String,
    @SerializedName("client_asr_text")
    val clientAsrText: String? = null,
    @SerializedName("asr_session_id")
    val asrSessionId: String? = null,
    @SerializedName("selected_text")
    val selectedText: String? = null,
    @SerializedName("clipboard_history")
    val clipboardHistory: List<String>? = null,
    val provider: String? = null,
    val model: String? = null,
    @SerializedName("fast_mode")
    val fastMode: Boolean = false,
    @SerializedName("transcript_language")
    val transcriptLanguage: String? = null
)

enum class AndroidQuickAction {
    @SerializedName("format")
    FORMAT,
    @SerializedName("polish")
    POLISH,
    @SerializedName("concise")
    CONCISE,
    @SerializedName("bullets")
    BULLETS
}

data class AndroidQuickActionRequest(
    val action: AndroidQuickAction,
    val text: String,
    @SerializedName("source_operation")
    val sourceOperation: String? = null
)

data class TextQuickActionResponse(
    val operation: String,
    @SerializedName("action_type")
    val actionType: ActionType = ActionType.PASTE,
    @SerializedName("input_text")
    val inputText: String,
    val result: String,
    @SerializedName("model_provider")
    val modelProvider: String? = null,
    @SerializedName("model_name")
    val modelName: String? = null,
    @SerializedName("credits_remaining")
    val creditsRemaining: Int? = null,
    val transcript: String? = null
)

// ==================== 错误处理 ====================

data class ApiErrorResponse(
    val code: String,
    val message: String,
    @SerializedName("config_update")
    val configUpdate: ConfigUpdatePayload? = null
)

sealed class ApiException(message: String) : Exception(message) {
    data class HttpError(
        val statusCode: Int,
        val errorResponse: ApiErrorResponse?
    ) : ApiException(errorResponse?.message ?: "未知错误")
    
    data class NetworkError(override val cause: Throwable) : ApiException("网络连接失败")
    data class DecodingError(override val cause: Throwable) : ApiException("数据解析失败")
    object Unauthorized : ApiException("登录已过期，请重新登录")
    object UserBanned : ApiException("账号已被禁用")
    object NotLoggedIn : ApiException("请先登录")
    
    val errorCode: String?
        get() = (this as? HttpError)?.errorResponse?.code
    
    val configUpdate: ConfigUpdatePayload?
        get() = (this as? HttpError)?.errorResponse?.configUpdate
    
    val isCreditsExhausted: Boolean
        get() = errorCode == "CREDITS_EXHAUSTED"
    
    val isNonRetryable: Boolean
        get() = errorCode == "DURATION_EXCEEDED"
}

/**
 * 面向 UI 的本地化错误文案:异常自带的 message 是简体中文常量,不能直接展示。
 * 固定类型错误按当前语言映射;HttpError 优先服务端 message;未知异常回退调用方给的本地化 fallback。
 */
fun Throwable.displayMessage(context: Context, fallback: String): String = when (this) {
    is ApiException.Unauthorized -> MobileStrings.sessionExpired(context)
    is ApiException.NotLoggedIn -> MobileStrings.loginRequired(context)
    is ApiException.UserBanned -> MobileStrings.userBanned(context)
    is ApiException.NetworkError -> MobileStrings.networkError(context)
    is ApiException.DecodingError -> fallback
    is ApiException.HttpError -> MobileStrings.apiErrorMessage(context, errorResponse?.code, errorResponse?.message ?: fallback)
    else -> message ?: fallback
}

// ==================== 录音配置 ====================

data class RecordingConfigResponse(
    @SerializedName("max_duration_sec")
    val maxDurationSec: Int
)

// ==================== 词典（热词） ====================

data class HotWordItem(
    val id: String,
    val word: String,
    @SerializedName("created_at")
    val createdAt: String?
)

data class HotWordCreateBody(
    val word: String
)

data class HotWordUpdateBody(
    val word: String
)

data class HotWordListResponse(
    val hotwords: List<HotWordItem>
)

// ==================== 人设配置 ====================

const val PERSONA_PROMPT_MAX_LENGTH = 4000
const val PERSONA_NAME_MAX_LENGTH = 60
const val PERSONA_DESC_MAX_LENGTH = 200
const val PERSONA_MAX_COUNT = 10

data class PersonaPrompts(
    @SerializedName("transcribe_prompt")
    val transcribePrompt: String? = null,
    @SerializedName("transcribe_enabled")
    val transcribeEnabled: Boolean = false,
    @SerializedName("rewrite_prompt")
    val rewritePrompt: String? = null,
    @SerializedName("rewrite_enabled")
    val rewriteEnabled: Boolean = false,
    @SerializedName("intent_hint")
    val intentHint: String? = null,
    @SerializedName("intent_enabled")
    val intentEnabled: Boolean = false
)

data class PersonaItem(
    val id: String,
    val name: String,
    val description: String? = null,
    @SerializedName("is_active")
    val isActive: Boolean = false,
    @SerializedName("is_builtin")
    val isBuiltin: Boolean = false,
    val prompts: PersonaPrompts = PersonaPrompts()
)

data class PersonaListResponse(
    val personas: List<PersonaItem>
)

data class PersonaCreateBody(
    val name: String,
    val description: String?,
    val prompts: PersonaPrompts
)

data class PersonaUpdateBody(
    val name: String?,
    val description: String?,
    val prompts: PersonaPrompts?
)

// ==================== 协议 ====================

data class AgreementItem(
    val id: String,
    val type: String,
    val title: String,
    val content: String,
    val version: String
)

data class AgreementsResponse(
    val agreements: List<AgreementItem>
)
